README.md

[![Arduino CI](https://github.com/RobTillaart/Radar/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/RADAR/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/RADAR/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/RADAR/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/RADAR/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/Radar/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/Radar.svg?maxAge=3600)](https://github.com/RobTillaart/Radar/releases)


# Radar

Arduino library for a pan tilt radar based upon 2 servo's.


## Warning

**WARNING: not extensively tested**
There are still a lot of things to be tested. 
The example code is now working to show the concept.


## Description

Radar is a concept library (not elaborated) that implements a Ping))) based Radar
with the use of two pan and tilt servo's. 
It was written after a question on the Arduino forum but never finished or tested it. 
Still, it has some interesting concepts with respect to how to determine the position 
of pan/tilt while the movements are in progress. 

The idea is that one knows how fast the pan and tilt servos are working in degrees per second.
From that number and the start moment one can determine approximately its position. 

Given its position while moving is interesting for radar purposes as one can determine e.g.
in which direction a ping is given and therefore which distance belongs to which pair of angles.

Note: no active development 


## Interface

TODO elaborate


### Constructor and config

- **RADAR(pan, tilt)** define pan / tilt pins of the radar. These should be PWM pins.

Note: no valid range checking or negative value check.
- **void setPanPerSecond(float pps)**
- **float getPanPerSecond()**
- **void setTiltPerSecond(float tps)**
- **float getTiltPerSecond()**


### basic moves

- **void gotoPan(const int16_t pan)**
- **int16_t getPan()**
- **void  gotoTilt(const int16_t tilt)**
- **int16_t getTilt()**
- **void gotoPanTilt(const int16_t pan, const int16_t tilt)**


### memory positions - store / recall?

- **uint8_t getMaxPositions()**
- **bool setPosition(const uint8_t index, const int16_t pan, const int16_t tilt)**
- **bool getPosition(const uint8_t index, int16_t & pan, int16_t & tilt)**
- **bool gotoPosition(const uint8_t index)**
- **void setHomePosition(const int16_t pan, const int16_t tilt)**
- **void gotoHomePosition()**


### feedback on positions.

- **bool isMoving()**
- **bool isPanMoving()**
- **bool isTiltMoving()**


## Operation

See examples


## Future

#### must
- improve documentation.

#### should
- test more (e.g. continuous servo versus 180 servo).
- add sketches.
- remember lastPing (angle pan tilt).
- check TODO's in code.
- move code from .h to .cpp

#### could
- single servo radar (Pan only) as derived class.
- dynamic allocation of position arrays.




CHANGELOG.md
# Change Log Radar

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.1.7] - 2022-11-23
- add changelog.md
- add RP2040 to build-CI  (commented as non AVR is not supported yet)
- fix keywords.txt
- minor edit readme.md
- minor edit unit test


## [0.1.6] - 2021-12-27
- update library.json
- update readme.md
- update license
- minor edits

## [0.1.5] - 2021-05-27
- fix Arduino-lint

## [0.1.4] - 2021-01-07
- Arduino-CI, unit-test

## [0.1.3] - 2020-07-06
- refactor
- clean up
- documentation

## [0.1.2] - 2017-07-16
- refactor & review

## [0.1.01] - 2015-03-06
- updated some code, still not functional

## [0.1.00] - 2013-09-30  (oldest version found ?)
- outline the idea



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2021-01-07
// PURPOSE: unit tests for the RADAR class
//          https://github.com/RobTillaart/RADAR
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual);               // a == b
// assertNotEqual(unwanted, actual);            // a != b
// assertComparativeEquivalent(expected, actual);    // abs(a - b) == 0 or (!(a > b) && !(a < b))
// assertComparativeNotEquivalent(unwanted, actual); // abs(a - b) > 0  or ((a > b) || (a < b))
// assertLess(upperBound, actual);              // a < b
// assertMore(lowerBound, actual);              // a > b
// assertLessOrEqual(upperBound, actual);       // a <= b
// assertMoreOrEqual(lowerBound, actual);       // a >= b
// assertTrue(actual);
// assertFalse(actual);
// assertNull(actual);

// // special cases for floats
// assertEqualFloat(expected, actual, epsilon);    // fabs(a - b) <= epsilon
// assertNotEqualFloat(unwanted, actual, epsilon); // fabs(a - b) >= epsilon
// assertInfinity(actual);                         // isinf(a)
// assertNotInfinity(actual);                      // !isinf(a)
// assertNAN(arg);                                 // isnan(a)
// assertNotNAN(arg);                              // !isnan(a)


#include <ArduinoUnitTests.h>


#include "radar.h" 


unittest_setup()
{
}


unittest_teardown()
{
}

/*
unittest(test_new_operator)
{
  assertEqualINF(exp(800));
  assertEqualINF(0.0/0.0);
  assertEqualINF(42);
  
  assertEqualNAN(INFINITY - INFINITY);
  assertEqualNAN(0.0/0.0);
  assertEqualNAN(42);
}
*/


unittest(test_constructor)
{
  fprintf(stderr, "VERSION: %s\n", (char *) RADAR_LIB_VERSION);

  RADAR radar(10, 11);

  assertEqual(10, radar.getMaxPositions());

  radar.setPanPerSecond(100);
  assertEqualFloat(100, radar.getPanPerSecond(), 0.0001);
  radar.setTiltPerSecond(100);
  assertEqualFloat(100, radar.getTiltPerSecond(), 0.0001);
}


unittest(test_positions)
{
  RADAR radar(10, 11);

  assertEqual(10, radar.getMaxPositions());

  for (int pos = 0; pos < 10; pos++)
  {
    fprintf(stderr, "%d\n", pos);
    int16_t x = random(360);
    int16_t y = random(180);
    assertTrue(radar.setPosition(pos, x, y));
    int16_t a, b;
    assertTrue(radar.getPosition(pos, a, b));
    assertEqual(x, a);
    assertEqual(y, b);
  }
}


unittest_main()

// --------


radar.h
#pragma once
//
//    FILE: radar.h
//  AUTHOR: Rob Tillaart
// VERSION: see RADAR_LIB_VERSION
// PURPOSE: Arduino library for a pan tilt radar.
//     URL: https://github.com/RobTillaart/RADAR


#include "Arduino.h"


#define RADAR_LIB_VERSION         (F("0.1.5"))


// TODO # positions in a begin() or constructor?
#ifndef RADAR_POSITIONS
#define RADAR_POSITIONS           10
#endif


class RADAR
{
public:
    RADAR(const uint8_t pinPan, const uint8_t pinTilt);

    // no valid range checking or negative value check.
    void     setPanPerSecond(float pps)  { _panPerSecond = pps; };
    float    getPanPerSecond()           { return _panPerSecond; };
    void     setTiltPerSecond(float tps) { _tiltPerSecond = tps; };
    float    getTiltPerSecond()          { return _tiltPerSecond; };

    // basic moves
    void     gotoPan(const int16_t pan);
    int16_t  getPan();
    void     gotoTilt(const int16_t tilt);
    int16_t  getTilt();
    void     gotoPanTilt(const int16_t pan, const int16_t tilt);

    // memory positions - store / recall?
    uint8_t  getMaxPositions()            { return RADAR_POSITIONS; };
    bool     setPosition(const uint8_t index, const int16_t pan, const int16_t tilt);
    bool     getPosition(const uint8_t index, int16_t & pan, int16_t & tilt);
    bool     gotoPosition(const uint8_t index);
    void     setHomePosition(const int16_t pan, const int16_t tilt);
    void     gotoHomePosition();

    // feedback on positions.
    bool     isMoving()     { return isPanMoving() || isTiltMoving(); };
    bool     isPanMoving()  { return getPan() != _pan; };
    bool     isTiltMoving() { return getTilt() != _tilt; };

    // TODO NIY
    uint32_t ping();
    uint32_t ping(const int16_t pan, const int16_t tilt);

private:
    int16_t  _pinPan;
    int16_t  _pinTilt;

    int16_t  _prevPan;
    int16_t  _pan;
    int16_t  _homePan;
    uint32_t _lastPanTime;

    int16_t  _prevTilt;
    int16_t  _tilt;
    int16_t  _homeTilt;
    uint32_t _lastTiltTime;

    int16_t  _panArray[RADAR_POSITIONS];
    int16_t  _tiltArray[RADAR_POSITIONS];

    float    _panPerSecond = 15;
    float    _tiltPerSecond = 15;
};


// -- END OF FILE --


radar.cpp
//
//    FILE: radar.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.5
// PURPOSE: Arduino library for a pan tilt radar.
//     URL: https://github.com/RobTillaart/RADAR
//
//  HISTORY
//  0.1.00  2013-09-30  (?) outline
//  0.1.01  2015-03-06  updated some code, still not functional
//  0.1.2   2017-07-16  refactor & review
//  0.1.3   2020-07-06  refactor, clean up and some documentation
//  0.1.4   2021-01-07  arduino-ci, unit-test
//  0.1.5   2021-05-27  fix arduino-lint


#include "radar.h"


////////////////////////////////////////////////////////////
//
// CONSTRUCTOR
//
RADAR::RADAR(const uint8_t pinPan, const uint8_t pinTilt)
{
  _pinPan = pinPan;
  _pinTilt = pinTilt;
  _homePan = 0;
  _homeTilt = 0;
  _pan = _prevPan = 0;
  _tilt = _prevTilt = 0;
  _panPerSecond = 15;
  _tiltPerSecond = 15;
  _lastPanTime = _lastTiltTime = millis();
  for (uint8_t i = 0; i < RADAR_POSITIONS; i++)
  {
    _panArray[i] = 0;
    _tiltArray[i] = 0;
  }
}


// PUBLIC
void RADAR::gotoPan(const int16_t pan)
{
  if (pan == _pan) return;
  _prevPan = getPan();
  _pan = pan;
  analogWrite(_pinPan, _pan);
  _lastPanTime = millis();
}


int16_t RADAR::getPan()
{
  // ESTIMATE current position on time it takes to go from previous to new
  if (_pan == _prevPan) return _pan;

  // if (enough time passed to move to new position) return new position
  uint32_t duration = millis() - _lastPanTime;
  uint32_t movement = round(duration * _panPerSecond * 0.001);
  if ( abs(_pan - _prevPan) <= movement) return _pan;

  // else estimate PAN by linear interpolation 
  if (_pan > _prevPan) return _prevPan + movement;
  return _prevPan - movement;
}


void RADAR::gotoTilt(const int16_t tilt)
{
  if (tilt == _tilt) return;

  _prevTilt = getTilt();
  _tilt = tilt;
  analogWrite(_pinTilt, _tilt);  // 0..180
  _lastTiltTime = millis();
}


int16_t RADAR::getTilt()
{
  // ESTIMATE current position on time it takes to go from previous to new
  if (_tilt == _prevTilt) return _tilt;

  // if (enough time passed to move to new position) return new position
  uint32_t duration = millis() - _lastTiltTime;
  uint32_t movement = round(duration * _tiltPerSecond * 0.001);
  if (abs(_tilt - _prevTilt) <= movement) return _tilt;

  // estimate TILT by linear interpolation 
  if (_tilt > _prevTilt) return _prevTilt + movement;  
  return _prevTilt - movement;
}


void RADAR::gotoPanTilt(const int16_t pan, const int16_t tilt)
{
  gotoPan(pan);
  gotoTilt(tilt);
}


bool RADAR::setPosition(const uint8_t index, const int16_t pan, const int16_t tilt)
{
  if (index >= RADAR_POSITIONS) return false;
  _panArray[index] = pan;
  _tiltArray[index] = tilt;
  return true;
}


bool RADAR::getPosition(const uint8_t index, int16_t & pan, int16_t & tilt)
{
  if (index >= RADAR_POSITIONS) return false;
  pan = _panArray[index];
  tilt = _tiltArray[index];
  return true;
}


bool RADAR::gotoPosition(const uint8_t index)
{
  if (index >= RADAR_POSITIONS) return false;
  gotoPan(_panArray[index]);
  gotoTilt(_tiltArray[index]);
  return true;
}


void RADAR::setHomePosition(const int16_t pan, const int16_t tilt)
{
  _homePan = pan;
  _homeTilt = tilt;
}


void RADAR::gotoHomePosition()
{
  gotoPan(_homePan);
  gotoTilt(_homeTilt);
}


uint32_t RADAR::ping()
{
  // TODO ping code here - playground or teckel's improved ping)))
  return 0;
}


uint32_t RADAR::ping(const int16_t pan, const int16_t tilt)
{
  gotoPan(pan);
  gotoTilt(tilt);
  while (isMoving());
  return ping();
}
// PRIVATE

// TODO distil private parts (getPan and getTilt share a lot


// -- END OF FILE --

